### PR TITLE
feat: integrate GFM-style alerts (#40) v0.12

### PR DESCRIPTION
**Pull request by mcdurdin**
* GFM-style alerts for keymanapp/help.keyman.com#954

### CODE REVIEW

**Comment by mcdurdin** [comment]
bug:

```
#0 {closure}() #1 spl_autoload_call() called at [/var/www/html/_common/MarkdownHost.php:90] #2 Keyman\Site\Common\MarkdownHost->__construct() called at [/var/www/html/_includes/md/mdhost.php:10] Unable to find class Keyman\Site\Common\GFMAlerts
```

**Comment by Meng-Heng** [comment]
@mcdurdin, do I need to test this locally with the alerts/v0.12 branch first?

**Comment by mcdurdin** [comment]
> @mcdurdin, do I need to test this locally with the alerts/v0.12 branch first?

I am running some tests locally now before merge. I think there may be some issues still, will let you know

**Comment by mcdurdin** [comment]
First issue resolved by adding GFMAlerts.php to bootstrap.inc.sh.

Second issue:

```
2024-11-08 09:01:38 [Fri Nov 08 02:01:38.403058 2024] [php7:error] [pid 18] [client 172.17.0.1:53252] PHP Fatal error:  Uncaught TypeError: Argument 1 passed to Parsedown::element() must be of the type array, string given, called in /var/www/vendor/erusev/parsedown/Parsedown.php on line 1545 and defined in /var/www/vendor/erusev/parsedown/Parsedown.php:1470\nStack trace:\n#0 /var/www/vendor/erusev/parsedown/Parsedown.php(1545): Parsedown->element()\n#1 /var/www/vendor/erusev/parsedown/Parsedown.php(1518): Parsedown->elements()\n#2 /var/www/vendor/erusev/parsedown/Parsedown.php(297): Parsedown->element()\n#3 /var/www/vendor/erusev/parsedown/Parsedown.php(39): Parsedown->lines()\n#4 /var/www/html/_common/MarkdownHost.php(105): Parsedown->text()\n#5 /var/www/html/_includes/md/mdhost.php(10): Keyman\\Site\\Common\\MarkdownHost->__construct()\n#6 {main}\n  thrown in /var/www/vendor/erusev/parsedown/Parsedown.php on line 1470
```

**Comment by mcdurdin** [comment]
Also I note we'll need to add CSS to help.keyman.com to support the look and feel of this

**Comment by Meng-Heng** [comment]
Yes, Marc. Also I need to tell you this. It does not support new line, and it must go with this format:
```
> [!WARNING] Testing my Alerts.

<!--  -->

> [!NOTE] Testing my Alerts.

<!--  -->

> [!TIPS] Testing my Alerts
```

image

File: _common/MarkdownHost.php [hunk 1]
```diff
@@ -95,12 +95,14 @@ function __construct($file) {
 
       // Performs the parsing + prettification of Markdown for display through PHP.
       $Parsedown = new \ParsedownExtra();
```
Comment (darcywong00) [inline _common/MarkdownHost.php:97]: Is $Parsedown still needed?

**Comment by darcywong00** [comment]
Do  we want to rebase this onto chore/v0.12?

File: _common/MarkdownHost.php (hunk 1 above)
Comment (Meng-Heng) [reply _common/MarkdownHost.php:97]: I think no, @darcywong00. I will remove it since nothing is breaking while testing.

**Comment by darcywong00** [comment]
Can we document the GFM-style markings?
Similar to https://github.com/keymanapp/help.keyman.com/issues/954#issuecomment-2092038059

Maybe it goes in https://github.com/keymanapp/shared-sites/blob/main/_common/README.md (add a new section on GFMAlert)

**Review by darcywong00 — DISMISSED** [review]
lgtm

**Comment by Meng-Heng** [comment]
I got the image and logos.
image

**Review by darcywong00 — CHANGES_REQUESTED** [review]
Is it possible to not require a comment between multiple GTM-style alerts?

Missing it results in these PHP errors
```php
Fatal error: Uncaught TypeError: Argument 1 passed to Parsedown::element() must be of the type array, string given, called in /var/www/vendor/erusev/parsedown/Parsedown.php on line 1545 and defined in /var/www/vendor/erusev/parsedown/Parsedown.php:1470 
Stack trace: 
#0 /var/www/vendor/erusev/parsedown/Parsedown.php(1545): Parsedown->element() 
#1 /var/www/vendor/erusev/parsedown/Parsedown.php(1518): Parsedown->elements() 
#2 /var/www/vendor/erusev/parsedown/Parsedown.php(297): Parsedown->element() 
#3 /var/www/vendor/erusev/parsedown/Parsedown.php(39): Parsedown->lines() 
#4 /var/www/html/_common/MarkdownHost.php(104): Parsedown->text() 
#5 /var/www/html/_includes/includes/md/mdhost.php(11): Keyman\Site\Common\MarkdownHost->__construct() 
#6 {main} thrown in /var/www/vendor/erusev/parsedown/Parsedown.php on line 1470
```

**Review by darcywong00 — APPROVED** [review]
lgtm

File: _common/GFMAlerts.php [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            'tips' => '<svg class="octicon octicon-light-bulb mr-2" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path fill="green" d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>',
```
Comment (mcdurdin) [inline _common/GFMAlerts.php:39]:
```suggestion
            'tip' => '<svg class="octicon octicon-light-bulb mr-2" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path fill="green" d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>',
```

File: _common/GFMAlerts.php [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    public $VERSION = '1.0';
+    
+    public function __construct()
+    {
+       $this->BlockTypes['>'][] = 'Alert';
+    }
+
+    protected function blockQuote($line)
+    {
+        if(!preg_match('/^\>\s*\[\!(IMPORTANT|TIPS|NOTE|WARNING|CAUTION)\]/i', $line['text'], $matches)) {
```
Comment (mcdurdin) [inline _common/GFMAlerts.php:28]:
```suggestion
        if(!preg_match('/^\>\s*\[\!(IMPORTANT|TIP|NOTE|WARNING|CAUTION)\]/i', $line['text'], $matches)) {
```

File: _common/GFMAlerts.php [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        return $block;
+    }
+
+    protected function blockQuoteContinue($line, array $block) {
+        if(!preg_match('/^\>\s(.*)/', $line['text'], $matches)) {
+            return null;
+        }
+        
+        // Recognize a new Alert
+        if(preg_match('/^\>\s*\[\!(IMPORTANT|TIPS|NOTE|WARNING|CAUTION)\]/i', $line['text'])) {
```
Comment (mcdurdin) [inline _common/GFMAlerts.php:74]:
```suggestion
        if(preg_match('/^\>\s*\[\!(IMPORTANT|TIP|NOTE|WARNING|CAUTION)\]/i', $line['text'])) {
```

**Review by mcdurdin — COMMENTED** [review]
@Meng-Heng FYI, moving from 'tips' to 'tip'.

For now I am adding the CSS only to the help.keyman.com site

**Pull request merged**